File: Substitution/C++/SubstitutionTest.cpp
#include <gtest/gtest.h>
#include "Substitution.h"

const std::string key = "bcdefghijklmnopqrstuvwxyza";

TEST(SubstitutionTest, EncodeSimpleLowercase)
{
    const substitution s("abc", key);
    EXPECT_EQ(s.encode(), "bcd");
}

TEST(SubstitutionTest, DecodeSimpleLowercase)
{
    const substitution s("abc", key);
    const std::string encoded = s.encode();
    EXPECT_EQ(s.decode(encoded), "abc");
}

TEST(SubstitutionTest, EncodeUppercase)
{
    const substitution s("XYZ", key);
    EXPECT_EQ(s.encode(), "yza");
}

TEST(SubstitutionTest, DecodeUppercase)
{
    const substitution s("XYZ", key);
    const std::string encoded = s.encode();
    EXPECT_EQ(s.decode(encoded), "xyz");
}

TEST(SubstitutionTest, EncodeWithSpacesAndSymbols)
{
    const substitution s("hello world!", key);
    EXPECT_EQ(s.encode(), "ifmmp xpsme!");
}

TEST(SubstitutionTest, DecodeWithSpacesAndSymbols)
{
    const substitution s("hello world!", key);
    const std::string encoded = s.encode();
    EXPECT_EQ(s.decode(encoded), "hello world!");
}

TEST(SubstitutionTest, EmptyString)
{
    const substitution s("", key);
    EXPECT_EQ(s.encode(), "");
    EXPECT_EQ(s.decode(""), "");
}

File: Substitution/Rust/substitution_test.rs
#[cfg(test)]
mod tests {
    use super::substitution::Substitution;

    #[test]
    fn test_basic_encoding() {
        let sub = Substitution::new("abc".to_string(), "zyxwvutsrqponmlkjihgfedcba".to_string());
        assert_eq!(sub.encode(), "zyx");
    }

    #[test]
    fn test_basic_decoding() {
        let sub = Substitution::new("abc".to_string(), "zyxwvutsrqponmlkjihgfedcba".to_string());
        let encoded = sub.encode();
        let decoded = sub.decode(&encoded);
        assert_eq!(decoded, "abc");
    }

    #[test]
    fn test_mixed_case_and_symbols() {
        let sub = Substitution::new("Hello, World!".to_string(), "phqgiumeaylnofdxjkrcvstzwb".to_string());
        assert_eq!(sub.encode(), "Eainn, Vncfk!");
    }

    #[test]
    fn test_empty_string() {
        let sub = Substitution::new("".to_string(), "abcdefghijklmnopqrstuvwxyz".to_string());
        assert_eq!(sub.encode(), "");
        assert_eq!(sub.decode(&"".to_string()), "");
    }

    #[test]
    fn test_encode_decode_reversibility() {
        let input = "The quick brown fox jumps over 13 lazy dogs.";
        let key = "mnbvcxzlkjhgfdsapoiuytrewq";
        let sub = Substitution::new(input.to_string(), key.to_string());
        let encoded = sub.encode();
        let decoded = sub.decode(&encoded);
        assert_eq!(decoded, input);
    }

    #[test]
    fn test_key_shorter_than_26_letters() {
        let sub = Substitution::new("abc".to_string(), "xyz".to_string());
        
        assert_eq!(sub.encode(), "xyz");
        assert_eq!(sub.decode(&"xyz".to_string()), "abc");
    }

    #[test]
    fn test_key_with_repeated_characters() {
        let sub = Substitution::new("abcabc".to_string(), "aaaaaaaaaaaaaaaaaaaaaaaaaa".to_string());
        
        let encoded = sub.encode();
        assert_eq!(encoded, "aaaaaa");
        let decoded = sub.decode(&encoded);
        
        assert_eq!(decoded, "aaaaaa");
    }
}
